test failure: test_streaming_ui_empirical — AssertionError (env pollution from test_anthropic_provider_stream)

## Failure

`deile/tests/might/streaming-ui/test_streaming_ui.py::test_streaming_ui_empirical`

## Error

```
deile/tests/might/streaming-ui/test_streaming_ui.py:163: in test_streaming_ui_empirical
    assert result["tool_starts"] >= 1, result
E   AssertionError: {'contract': {'duration_ok': True, 'had_text_before_first_tool': False,
    'no_error': False, 'tool_round_trip_complete': False}, 'duration_s': 0.209,
    'events': 13, ...}
E   assert 0 >= 1
```

## Root cause

`deile/tests/core/models/test_anthropic_provider_stream.py:40` calls:

```python
os.environ.setdefault("ANTHROPIC_API_KEY", "sk-test-fake")
```

inside `_make_provider()`, which is invoked by every test in that module. Unlike `monkeypatch.setenv()`, this call **never restores the environment**. After those tests run, `ANTHROPIC_API_KEY` is permanently set to `"sk-test-fake"` for the rest of the process.

`test_streaming_ui_empirical` has a correct skip guard:

```python
if MODEL_KEY.startswith("anthropic:") and not os.getenv("ANTHROPIC_API_KEY"):
    return "ANTHROPIC_API_KEY not set"
```

But with the leaked fake key, `os.getenv("ANTHROPIC_API_KEY")` returns `"sk-test-fake"` (truthy), so the guard passes, the test runs, the Anthropic SDK rejects the fake key, the agent errors out, and `tool_starts` stays 0.

**Verified**: the test skips correctly when run in isolation (`pytest deile/tests/might/streaming-ui/test_streaming_ui.py -v`) but fails in the full suite (`pytest deile/tests/ -v`).

## Reproduce locally

```bash
# Fails: test_anthropic_provider_stream runs first and leaks the key
pytest deile/tests/ -v -k "test_text_stream_only or test_streaming_ui_empirical"

# Passes/skips: run in isolation (no prior pollution)
pytest deile/tests/might/streaming-ui/test_streaming_ui.py -v
```

## Environment

- Python: Python 3.11.15
- Commit: $(git rev-parse HEAD)
- Branch: main

## Suggested fix (one paragraph, specific)

Replace the bare `os.environ.setdefault(...)` in `_make_provider()` at `deile/tests/core/models/test_anthropic_provider_stream.py:40` with a `monkeypatch`-based approach: convert `_make_provider()` into a pytest fixture that accepts `monkeypatch` and calls `monkeypatch.setenv("ANTHROPIC_API_KEY", "sk-test-fake")`. All call sites (`test_text_stream_only`, `test_tool_use_stream_emits_lifecycle`, etc.) already receive the provider via a fixture pattern, so adding `monkeypatch` as a dependency to the provider fixture is a one-line change that guarantees automatic teardown and eliminates the cross-test pollution.

<!-- DEILE-FP-b629f17f0ec87d0f -->

--- a/deile/tests/core/models/test_anthropic_provider_stream.py
+++ b/deile/tests/core/models/test_anthropic_provider_stream.py
@@ -6,7 +6,6 @@ mapping, not network behaviour.
 
 from __future__ import annotations
 
-import os
 from contextlib import asynccontextmanager
 from types import SimpleNamespace
 from typing import List
@@ -18,7 +17,7 @@ from deile.core.models.base import ModelMessage
 from deile.core.models.stream_events import StreamEventType
 
 
-def _make_provider():
+def _make_provider(monkeypatch):
     from deile.core.models.anthropic_provider import AnthropicProvider
     from deile.core.models.catalog import ModelHandle, ModelPricing
     from deile.core.models.tier import ModelTier
@@ -37,8 +36,13 @@ def _make_provider():
     cfg = ProviderConfig(
         provider_id="anthropic", api_key_env="ANTHROPIC_API_KEY", base_url=None
     )
-    os.environ.setdefault("ANTHROPIC_API_KEY", "sk-test-fake")
+    monkeypatch.setenv("ANTHROPIC_API_KEY", "sk-test-fake")
     return AnthropicProvider(model_handle=handle, provider_config=cfg)
+
+
+@pytest.fixture
+def provider(monkeypatch):
+    return _make_provider(monkeypatch)
 
 
 def _e(etype, **kw):
@@ -66,8 +70,7 @@ async def _fake_stream(events: List, final_message):
 
 
 @pytest.mark.asyncio
-async def test_text_stream_only():
-    provider = _make_provider()
+async def test_text_stream_only(provider):
     final = SimpleNamespace(
         usage=SimpleNamespace(
             input_tokens=10,
@@ -103,8 +106,7 @@ async def test_text_stream_only():
 
 
 @pytest.mark.asyncio
-async def test_tool_use_stream_emits_lifecycle():
-    provider = _make_provider()
+async def test_tool_use_stream_emits_lifecycle(provider):
     final = SimpleNamespace(
         usage=SimpleNamespace(
             input_tokens=12,
@@ -152,8 +154,7 @@ async def test_tool_use_stream_emits_lifecycle():
     assert end_evt.arguments == {"command": "ls"}
 
 
-def test_format_assistant_tool_use_message_structure():
-    provider = _make_provider()
+def test_format_assistant_tool_use_message_structure(provider):
     msg = provider.format_assistant_tool_use_message(
         [("t1", "list_files", {"path": "."})], text_so_far="ok"
     )
@@ -165,8 +166,7 @@ def test_format_assistant_tool_use_message_structure():
     assert blocks[1]["input"] == {"path": "."}
 
 
-def test_format_tool_result_message_round_trip():
-    provider = _make_provider()
+def test_format_tool_result_message_round_trip(provider):
     msg = provider.format_tool_result_message("t1", "list_files", {"items": ["a.py"]})
     block = msg.metadata["_anthropic_content_blocks"][0]
     assert block["type"] == "tool_result"

--- a/deile/tests/core/models/test_gemini_provider_stream.py
+++ b/deile/tests/core/models/test_gemini_provider_stream.py
@@ -7,7 +7,6 @@ single TEXT_DELTA + TOOL_USE_START/END pair per round-trip.
 
 from __future__ import annotations
 
-import os
 from types import SimpleNamespace
 from unittest.mock import AsyncMock, MagicMock, patch
 
@@ -15,7 +14,7 @@ import pytest
 
 
 @pytest.fixture
-def gemini_provider():
+def gemini_provider(monkeypatch):
     """Construct a GeminiProvider without hitting the network."""
     from deile.core.models.gemini_provider import GeminiProvider
     from deile.core.models.catalog import ModelHandle, ModelPricing
@@ -35,7 +34,7 @@ def gemini_provider():
     cfg = ProviderConfig(
         provider_id="gemini", api_key_env="GOOGLE_API_KEY", base_url=None
     )
-    os.environ.setdefault("GOOGLE_API_KEY", "fake")
+    monkeypatch.setenv("GOOGLE_API_KEY", "fake")
     with patch("deile.core.models.gemini_provider.genai"):
         provider = GeminiProvider(handle, cfg)
     return provider

--- a/deile/tests/core/models/test_openai_provider_stream.py
+++ b/deile/tests/core/models/test_openai_provider_stream.py
@@ -3,7 +3,6 @@
 from __future__ import annotations
 
 import json
-import os
 from types import SimpleNamespace
 from typing import List
 from unittest.mock import patch
@@ -14,7 +13,7 @@ from deile.core.models.base import ModelMessage
 from deile.core.models.stream_events import StreamEventType
 
 
-def _make_provider(provider_cls_name: str = "OpenAIProvider"):
+def _make_provider(monkeypatch, provider_cls_name: str = "OpenAIProvider"):
     from deile.core.models.openai_provider import OpenAIProvider
     from deile.core.models.deepseek_provider import DeepSeekProvider
     from deile.core.models.catalog import ModelHandle, ModelPricing
@@ -34,7 +33,7 @@ def _make_provider(provider_cls_name: str = "OpenAIProvider"):
     cfg = ProviderConfig(
         provider_id="openai", api_key_env="OPENAI_API_KEY", base_url=None
     )
-    os.environ.setdefault("OPENAI_API_KEY", "sk-test-fake")
+    monkeypatch.setenv("OPENAI_API_KEY", "sk-test-fake")
     cls = OpenAIProvider if provider_cls_name == "OpenAIProvider" else DeepSeekProvider
     if cls is DeepSeekProvider:
         cfg = ProviderConfig(
@@ -42,8 +41,13 @@ def _make_provider(provider_cls_name: str = "OpenAIProvider"):
             api_key_env="DEEPSEEK_API_KEY",
             base_url="https://api.deepseek.com/v1",
         )
-        os.environ.setdefault("DEEPSEEK_API_KEY", "sk-test-fake")
+        monkeypatch.setenv("DEEPSEEK_API_KEY", "sk-test-fake")
     return cls(model_handle=handle, provider_config=cfg)
+
+
+@pytest.fixture
+def provider(monkeypatch):
+    return _make_provider(monkeypatch)
 
 
 async def _replay(chunks: List):
@@ -77,8 +81,7 @@ def _chunk_finish_with_usage(prompt_tokens=10, completion_tokens=2):
 
 
 @pytest.mark.asyncio
-async def test_text_stream_emits_text_and_usage():
-    provider = _make_provider()
+async def test_text_stream_emits_text_and_usage(provider):
     chunks = [
         _chunk_text("hello "),
         _chunk_text("world"),
@@ -103,9 +106,7 @@ async def test_text_stream_emits_text_and_usage():
 
 
 @pytest.mark.asyncio
-async def test_tool_call_stream_emits_lifecycle():
-    provider = _make_provider()
-
+async def test_tool_call_stream_emits_lifecycle(provider):
     def _delta_tool_call(index, id_=None, name=None, arguments=None):
         return SimpleNamespace(
             choices=[
@@ -164,8 +165,7 @@ async def test_tool_call_stream_emits_lifecycle():
     assert end_idx == start_idx + 1, "TOOL_USE_END must immediately follow TOOL_USE_START"
 
 
-def test_format_assistant_tool_use_message_structure():
-    provider = _make_provider()
+def test_format_assistant_tool_use_message_structure(provider):
     msg = provider.format_assistant_tool_use_message(
         [("c1", "echo", {"x": 1})], text_so_far=""
     )
@@ -175,8 +175,7 @@ def test_format_assistant_tool_use_message_structure():
     assert json.loads(tcs[0]["function"]["arguments"]) == {"x": 1}
 
 
-def test_format_tool_result_message_round_trip():
-    provider = _make_provider()
+def test_format_tool_result_message_round_trip(provider):
     msg = provider.format_tool_result_message("c1", "echo", {"ok": True})
     tr = msg.metadata["_openai_tool_result"]
     assert tr["tool_call_id"] == "c1"
@@ -218,11 +217,10 @@ def _chunk_with_reasoning(reasoning: str, content: str = ""):
 
 
 @pytest.mark.asyncio
-async def test_reasoning_content_emitted_in_usage_final_for_stop_path():
+async def test_reasoning_content_emitted_in_usage_final_for_stop_path(provider):
     """When a provider streams delta.reasoning_content and finishes with 'stop',
     the accumulated text must appear in the USAGE_FINAL event's reasoning_content
     field so agent.py can store it for the next turn."""
-    provider = _make_provider()
     chunks = [
         _chunk_with_reasoning("<think>step1</think>"),
         _chunk_with_reasoning("<think>step2</think>", content="Hi"),
@@ -244,12 +242,10 @@ async def test_reasoning_content_emitted_in_usage_final_for_stop_path():
 
 
 @pytest.mark.asyncio
-async def test_reasoning_content_emitted_in_tool_use_end():
+async def test_reasoning_content_emitted_in_tool_use_end(provider):
     """When finish_reason=='tool_calls', reasoning_content must be attached to
     every TOOL_USE_END event so ToolLoopExecutor can pass it to
     format_assistant_tool_use_message for the next iteration."""
-    provider = _make_provider()
-
     def _tc_delta(index, id_=None, name=None, arguments=None, reasoning=None):
         return SimpleNamespace(
             choices=[
@@ -298,10 +294,9 @@ async def test_reasoning_content_emitted_in_tool_use_end():
     assert end_evt.reasoning_content == "<think>I need a tool</think>"
 
 
-def test_format_assistant_tool_use_message_preserves_reasoning_content():
+def test_format_assistant_tool_use_message_preserves_reasoning_content(provider):
     """reasoning_content kwarg must be stored in metadata so _to_openai_messages
     can echo it back to the API on the next round-trip."""
-    provider = _make_provider()
     msg = provider.format_assistant_tool_use_message(
         [("c1", "echo", {"x": 1})],
         text_so_far="",

